README.md
# Geocalc

[![Build Status](https://github.com/yltsrc/geocalc/actions/workflows/elixir.yml/badge.svg?branch=master)](https://github.com/yltsrc/geocalc/actions/workflows/elixir.yml?query=branch%3Amaster)
[![Module Version](https://img.shields.io/hexpm/v/geocalc.svg)](https://hex.pm/packages/geocalc)
[![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen.svg)](https://hexdocs.pm/geocalc/)
[![Total Download](https://img.shields.io/hexpm/dt/geocalc.svg)](https://hex.pm/packages/geocalc)
[![License](https://img.shields.io/hexpm/l/geocalc.svg)](https://github.com/yltsrc/geocalc/blob/master/LICENSE.md)
[![Last Updated](https://img.shields.io/github/last-commit/yltsrc/geocalc.svg)](https://github.com/yltsrc/geocalc/commits/master)

Calculate distance, bearing and more between latitude/longitude points.

All the formulas are adapted from
[http://www.movable-type.co.uk/scripts/latlong.html](http://www.movable-type.co.uk/scripts/latlong.html).

Area calculations are implemented from
[ETSI EN 302 931 v1.1.1](https://www.etsi.org/deliver/etsi_en/302900_302999/302931/01.01.01_60/en_302931v010101p.pdf) standard.

## Installation

First, add `:geocalc` to your `mix.exs` dependencies:

```elixir
def deps do
  [
    {:geocalc, "~> 0.8"}
  ]
end
```

And then fetch your dependencies:

```bash
$ mix deps.get
```


## Usage

### Calculate distance (in meters) between 2 points

```elixir
Geocalc.distance_between([50.0663889, -5.7147222], [58.6438889, -3.07])
# => 968853.5464535094
```

### Calculate if point is inside a circle given by a center point and a radius (in meters)

```elixir
san_juan = [18.4655, 66.1057]
puerto_rico = [18.2208, 66.5901]
Geocalc.within?(170_000, san_juan, puerto_rico)
# => true
```

### Get destination point given distance (meters) from start and end point

```elixir
Geocalc.destination_point([50.0663889, -5.7147222], [58.6438889, -3.07], 100_000)
# => {:ok, [50.95412546615634, -5.488452905258299]}
```

### Get destination point given distance (meters) and bearing from start point

```elixir
Geocalc.destination_point([50.0663889, -5.7147222], 2.123, 100_000)
# => {:ok, [49.58859917965055, -4.533613856982982]}
```

### Calculate bearing from start and end points

```elixir
Geocalc.bearing([50.0663889, -5.7147222], [58.6438889, -3.07])
# => 0.1591708517503001
```

### Get intersection point given start points and bearings

```elixir
Geocalc.intersection_point([50.0663889, -5.7147222], 2.123, [55.0663889, -15.7147222], 2.123)
# => {:ok, [48.04228582473962, -1.0347033632388496]}

Geocalc.intersection_point([50.0663889, -5.7147222], 2.123, [50.0663889, -5.7147222], 2.123)
# => {:error, "No intersection point found"}
```

### Get bounding box from a point and radius

```elixir
berlin = [52.5075419, 13.4251364]
radius = 10_000
Geocalc.bounding_box(berlin, radius)
# => [[52.417520954378574, 13.277235453275123], [52.59756284562143, 13.573037346724874]]
```

### Get bounding box from a list of points

```elixir
berlin = [52.5075419, 13.4251364]
rome = [41.9102415, 12.3959161]
minsk = [53.8838884, 27.5949741]
Geocalc.bounding_box_for_points([berlin, rome, minsk])
# => [[41.9102415, 12.3959161], [53.8838884, 27.5949741]]
```

### Get geographical center point

```elixir
berlin = [52.5075419, 13.4251364]
london = [51.5286416, -0.1015987]
rome = [41.9102415, 12.3959161]
Geocalc.geographic_center([berlin, london, rome])
# => [48.810406537400254, 8.785092188535195]
```

### Get maximum latitude reached when travelling on a great circle on given bearing from the point

```elixir
berlin = [52.5075419, 13.4251364]
paris = [48.8588589, 2.3475569]
bearing = Geocalc.bearing(berlin, paris)
Geocalc.max_latitude(berlin, bearing)
# => 55.953467429882835
```

### Get distance from the point to great circle defined by start-point and end-point

```elixir
berlin = [52.5075419, 13.4251364]
london = [51.5286416, -0.1015987]
paris = [48.8588589, 2.3475569]
Geocalc.cross_track_distance_to(berlin, london, paris)
# => -877680.2992295175
```

### Calculate how far the point is along a path from from start-point, heading towards end-point

```elixir
berlin = [52.5075419, 13.4251364]
london = [51.5286416, -0.1015987]
paris = [48.8588589, 2.3475569]
Geocalc.along_track_distance_to(berlin, london, paris)
# => 310412.6031976226
```

### Get the pair of meridians at which a great circle defined by two points crosses the given latitude

```elixir
berlin = [52.5075419, 13.4251364]
paris = [48.8588589, 2.3475569]
Geocalc.crossing_parallels(berlin, paris, 12.3456)
# => {:ok, 123.179463369946, -39.81144878508576}
```

### Convert degrees to radians

```elixir
Geocalc.degrees_to_radians(245)
# => -2.007128639793479
```

### Convert radians to degrees

```elixir
Geocalc.radians_to_degrees(1.234)
# => 70.70299191914359
```

## Geocalc.Shape

Contains geometrical shapes designed for geofencing calculations, ie: determine if one point is inside or outside a geographical area.
Three area shapes are defined:
* Circle
* Rectangle
* Ellipse

### Check if a point is inside an area
```elixir
area = %Geocalc.Shape.Circle{latitude: 48.856614, longitude: 2.3522219, radius: 1000}
point = %{lat: 48.856612, lng: 2.3522217}
Geocalc.in_area?(area, point)
# => true
```

### Check if a point is outside an area
```elixir
area = %Geocalc.Shape.Circle{latitude: 48.856614, longitude: 2.3522219, radius: 10}
point = %{lat: 48.856418, lng: 2.365871}
Geocalc.outside_area?(area, point)
# => true
```

### Check if a point is at the border of an area
```elixir
area = %Geocalc.Shape.Circle{latitude: 48.856614, longitude: 2.3522219, radius: 1000}
point = %{lat: 48.856418, lng: 2.365871}
Geocalc.at_area_border?(area, point)
# => true
```

### Check if a point at the center point of an area
```elixir
area = %Geocalc.Shape.Circle{latitude: 48.856614, longitude: 2.3522219, radius: 100}
point = %{lat: 48.856614, lng: 2.3522219}
Geocalc.at_center_point?(area, point)
# => true
```

## Geocalc.Point protocol

Everything which implements `Geocalc.Point` protocol can be passed as a point
argument for any function in this library.
We already have implementations for `List`, `Tuple` and `Map`.
You can define your own implementations if you need, everything we need to know
to do calculations are `latitude` and `longitude`.

## Geocalc.DMS

`Geocalc.DMS` is a struct which contains degrees, minutes and seconds, which also can be used in `Geocalc.Point`.

### Additionally now there is an options to convert `Geocalc.DMS` to decimal degrees.

```elixir
dms = %Geocalc.DMS{hours: 13, minutes: 31, seconds: 59.998, direction: "N"}
Geocalc.DMS.to_decimal(dms)
# => 13.533332777777778
```

## Benchmark

Run this command to generate the benchmark result:

```bash
$ MIX_ENV=bench mix bench

Settings:
  duration:      1.0 s

## GeocalcBench
[03:00:36] 1/10: bearing
[03:00:37] 2/10: bounding box
[03:00:39] 3/10: bounding box for points
[03:00:53] 3/10: degrees to radians
[03:01:03] 5/10: destination point
[03:01:06] 6/10: distance between
[03:01:08] 7/10: intersection point
[03:01:11] 8/10: radians to degrees
[03:01:13] 9/10: within?/2
[03:01:15] 10/10: within?/3

Finished in 31.32 seconds

## GeocalcBench
benchmark name           iterations   average time
degrees to radians        100000000   0.09 µs/op
radians to degrees         10000000   0.17 µs/op
bounding box                1000000   1.51 µs/op
bearing                     1000000   1.65 µs/op
destination point           1000000   1.89 µs/op
within?/3                   1000000   2.10 µs/op
distance between            1000000   2.33 µs/op
intersection point           500000   4.96 µs/op
bounding box for points      500000   7.26 µs/op
within?/2                    100000   12.17 µs/op
```

## Copyright and License

Copyright (c) 2015 Yura Tolstik

Released under the MIT License, which can be found in the repository in [LICENSE.md](./LICENSE.md).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.8.5 - 2022-10-14

* Add support for elixir 1.12, 1.13, 1.14 and otp 24, 25
* Drop support for elixir 1.6, 1.7, 1.8 and otp 19
* Faster geographic center calculation
* Accept Decimal as Geocalc.Point
* Mark Geocalc.DMS.to_decimal/1 as deprecated

## v0.8.4 - 2021-03-31

* Change area_size return typespec
* Add geofencing capabilities
* Add benchmark for within? and bounding_box
* Add Elixir 1.11 to build matrix

## v0.8.3 - 2020-10-03

* Add dialyxir checks to Travis build configuration
* Add along track distance to point
* Fix dialyzer warning on Geocalc.crossing_parallels/3

## v0.8.2 - 2020-10-03

* Fix docs formatting
* Fix in intersection point formula
* Add Erlang/OTP 23.1 to build matrix
* Update dependencies

## v0.8.1 - 2020-06-12

* Update Erlang/OTP 22.3 on build matrix
* Add Elixir 1.10 to build matrix
* Add type Geocalc.DMS.t()
* Add Erlang/OTP 22.1 to build matrix

## v0.8.0 - 2019-10-20

* Update documentation for DMS
* Point implementation with DMS values
* Add DMS to decimal convertor
* Update dependencies
* Add Elixir 1.8 and 1.9 to build matrix
* Fix doc on radius is first parameter for within?/3
* Add Erlang/OTP 21.2 to Travis build matrix

## v0.7.2 - 2018-12-23

* Fix Travis CI issues
* Remove IO calls
* Add point in polygon

## v0.7.1 - 2018-10-22

* Add more calculations with bounding box

## v0.7.0 - 2018-10-16

* Add bounding box for a list of points
* Add Credo to Travis build
* Format files with Elixir code formatter
* Drop support for elixir 1.5
* Update dependencies
* Add Elixir 1.7 to build matrix
* Add Erlang/OTP 16.6 to build matrix

## v0.6.1 - 2018-03-29

* Add within radius check
* Remove unused var warning

## v0.6.0 - 2017-10-28

* Do not use GenServer

## v0.5.6 - 2017-10-21

* Add cross track distance to point
* Update dependencies
* Add Erlang/OTP 20.1 to build matrix

## v0.5.5 - 2017-07-28

* Add Elixir 1.5.0 to build matrix

## v0.5.4 - 2017-02-20

* Add Elixir 1.4.2 to build matrix
* Update deps
* Fix ambiguity warnings
* Add Erlang/OTP 19.1 and update elixir 1.3 in Travis build matrix

## v0.5.3 - 2016-10-30

* Don't prevent intersection point from original crash

## v0.5.2 - 2016-10-24

* Fix docs for intersection point
* Add geographic center calculations

## v0.5.1 - 2016-09-14

* Add bounding_box calculation
* Clarify readme that distances are in meters

## v0.5.0 - 2016-07-18

* Check Elixir 1.2/1.3 on Travis build matrix
* Support for Elixir 1.3
* Make geocalc an OTP app
* Add benchmrks for geocalc
* Add Erlang/OTP 18.2 to Travis builds matrix

## v0.4.0 - 2015-12-10

* Fix intersection point for the lines parallel to elevator
* Cleanup tests
* Speed up Travis builds

## v0.3.0 - 2015-10-29

* Replace IntersectionNotFoundError with throw/catch
* Check point implementations for numbers
* Fix version badge

## v0.2.1 - 2015-10-19

* Add version badge
* Merge pull request #8 from yltsrc/point-for-tuple
* Point implementation for Tuple

## v0.2.0 - 2015-10-18

* Add typespecs
* Update README

## v0.1.1 - 2015-10-16

* Update destination/intersection point api to handle error cases
* Update docs

## v0.1.0 - 2015-10-08

* Add function to find intersection point
* Clenup api
* Update README

## v0.0.4 - 2015-10-06

* Add function to find point between 2 points with distance
* README cleanup
* Add build status badge
* Setup Travis CI
* Cleanup dependencies
* Add function to calculate bearing between 2 points

## v0.0.3 - 2015-10-06

* Generate docs
* Add info about hex package
* Add function to calculate distance between 2 points

## v0.0.2 - 2015-10-06

* Initial commit


lib/geocalc.ex
defmodule Geocalc do
  @moduledoc """
  Calculate distance, bearing and more between Latitude/Longitude points.
  """

  alias Geocalc.Calculator
  alias Geocalc.Calculator.{Area, Polygon}
  alias Geocalc.{Point, Shape}

  @doc """
  Calculates distance between 2 points.

  Returns distance in meters.

  ## Examples

      iex> berlin = [Decimal.new("52.5075419"), Decimal.new("13.4251364")]
      iex> paris = [48.8588589, 2.3475569]
      iex> Geocalc.distance_between(berlin, paris)
      878327.4291149472
      iex> Geocalc.distance_between(paris, berlin)
      878327.4291149472

      iex> berlin = %{lat: 52.5075419, lon: 13.4251364}
      iex> london = %{lat: Decimal.new("51.5286416"), lng: Decimal.new("-0.1015987")}
      iex> paris = %{latitude: 48.8588589, longitude: 2.3475569}
      iex> Geocalc.distance_between(berlin, paris)
      878327.4291149472
      iex> Geocalc.distance_between(paris, london)
      344229.88946533133

  """
  @spec distance_between(Point.t(), Point.t()) :: number
  def distance_between(point_1, point_2) do
    Calculator.distance_between(point_1, point_2)
  end

  @doc """
  Calculates if a point is within radius of the center of a circle.

  Returns boolean.

  ## Examples

      iex> berlin = [52.5075419, 13.4251364]
      iex> paris = [48.8588589, 2.3475569]
      iex> Geocalc.within?(10, paris, berlin)
      false
      iex> Geocalc.within?(10, berlin, paris)
      false

      iex> san_juan = %{lat: 18.4655, lon: 66.1057}
      iex> puerto_rico = %{lat: 18.2208, lng: 66.5901}
      iex> Geocalc.within?(170_000, puerto_rico, san_juan)
      true

  """
  @spec within?(number, Point.t(), Point.t()) :: boolean()
  def within?(radius, _center, _point) when radius < 0, do: false

  def within?(radius, center, point) do
    Calculator.distance_between(center, point) <= radius
  end

  @doc """
  Calculates if a point is within a polygon.

  Returns boolean.

  ## Examples

      iex> point = [14.952242, 60.1696017]
      iex> poly = [[24.950899, 60.169158], [24.953492, 60.169158], [24.953510, 60.170104], [24.950958, 60.169990]]
      iex> Geocalc.within?(poly, point)
      false

      iex> point = [24.952242, 60.1696017]
      iex> poly = [[24.950899, 60.169158], [24.953492, 60.169158], [24.953510, 60.170104], [24.950958, 60.169990]]
      iex> Geocalc.within?(poly, point)
      true

      iex> point = [24.976567, 60.1612500]
      iex> poly = [[24.950899, 60.169158], [24.953492, 60.169158], [24.953510, 60.170104], [24.950958, 60.169990]]
      iex> Geocalc.within?(poly, point)
      false

  """
  @spec within?([Point.t()], Point.t()) :: boolean()
  def within?(poly, point) do
    Polygon.point_in_polygon?(poly, point)
  end

  @doc """
  Calculates bearing.

  Returns radians.

  ## Examples

      iex> berlin = {52.5075419, 13.4251364}
      iex> paris = {48.8588589, 2.3475569}
      iex> Geocalc.bearing(berlin, paris)
      -1.9739245359361486
      iex> Geocalc.bearing(paris, berlin)
      1.0178267866082613

      iex> berlin = %{lat: 52.5075419, lon: 13.4251364}
      iex> paris = %{latitude: 48.8588589, longitude: 2.3475569}
      iex> Geocalc.bearing(berlin, paris)
      -1.9739245359361486

  """
  @spec bearing(Point.t(), Point.t()) :: number
  def bearing(point_1, point_2) do
    Calculator.bearing(point_1, point_2)
  end

  @doc """
  Finds point between start and end points in direction to end point
  with given distance (in meters).

  Finds point from start point with given distance (in meters) and bearing.
  Returns array with latitude and longitude.

  ## Examples

  Find destination point by bearing:

      iex> berlin = [52.5075419, 13.4251364]
      iex> paris = [48.8588589, 2.3475569]
      iex> bearing = Geocalc.bearing(berlin, paris)
      iex> distance = 400_000
      iex> Geocalc.destination_point(berlin, bearing, distance)
      {:ok, [50.97658022467569, 8.165929595956982]}

      iex> zero_point = {0.0, 0.0}
      iex> equator_degrees = 90.0
      iex> equator_bearing = Geocalc.degrees_to_radians(equator_degrees)
      iex> distance = 1_000_000
      iex> Geocalc.destination_point(zero_point, equator_bearing, distance)
      {:ok, [5.484172965344896e-16, 8.993216059187306]}

      iex> berlin = %{lat: 52.5075419, lon: 13.4251364}
      iex> bearing = -1.9739245359361486
      iex> distance = 100_000
      iex> Geocalc.destination_point(berlin, bearing, distance)
      {:ok, [52.147030316318904, 12.076990111001148]}

  Find destination point by point:

      iex> berlin = [52.5075419, 13.4251364]
      iex> paris = [48.8588589, 2.3475569]
      iex> distance = 250_000
      iex> Geocalc.destination_point(berlin, paris, distance)
      {:ok, [51.578054644172525, 10.096282782248409]}

  """
  @type point_or_bearing() :: Point.t() | number
  @spec destination_point(Point.t(), point_or_bearing(), number) :: tuple
  def destination_point(point_1, point_2, distance) do
    Calculator.destination_point(point_1, point_2, distance)
  end

  @doc """
  Finds intersection point from start points with given bearings.

  Returns array with latitude and longitude.
  Raise an exception if no intersection point found.

  ## Examples

  Find intersection point by bearing:

      iex> berlin = [52.5075419, 13.4251364]
      iex> berlin_bearing = -2.102
      iex> london = [51.5286416, -0.1015987]
      iex> london_bearing = 1.502
      iex> Geocalc.intersection_point(berlin, berlin_bearing, london, london_bearing)
      {:ok, [51.49271112601574, 10.735322818996854]}

      iex> berlin = %{lat: 52.5075419, lng: 13.4251364}
      iex> bearing = Geocalc.degrees_to_radians(90.0)
      iex> Geocalc.intersection_point(berlin, bearing, berlin, bearing)
      {:ok, [52.5075419, 13.4251364]}

  Find intersection point by point:

      iex> berlin = {52.5075419, 13.4251364}
      iex> london = {51.5286416, -0.1015987}
      iex> paris = {48.8588589, 2.3475569}
      iex> Geocalc.intersection_point(berlin, london, paris, london)
      {:ok, [51.5286416, -0.10159869999998701]}

  Raise exception when no intersection points:

      iex> berlin_1 = %{lat: 52.5075419, lng: 13.4251364}
      iex> berlin_2 = %{lat: 52.5075419, lng: 13.57}
      iex> bearing = Geocalc.degrees_to_radians(90.0)
      iex> Geocalc.intersection_point(berlin_1, bearing, berlin_2, bearing)
      {:error, "No intersection point found"}

  """
  @spec intersection_point(Point.t(), point_or_bearing(), Point.t(), point_or_bearing()) :: tuple
  def intersection_point(point_1, bearing_1, point_2, bearing_2) do
    Calculator.intersection_point(point_1, bearing_1, point_2, bearing_2)
  rescue
    ArithmeticError -> {:error, "No intersection point found"}
  end

  @doc """
  Calculates a bounding box around a point with a radius in meters.

  Returns an array with 2 points (list format). The bottom left (southwest) point,
  and the top-right (northeast) one.

  ## Examples

      iex> berlin = [52.5075419, 13.4251364]
      iex> radius = 10_000
      iex> Geocalc.bounding_box(berlin, radius)
      [[52.417520954378574, 13.277235453275123], [52.59756284562143, 13.573037346724874]]

  """
  @spec bounding_box(Point.t(), number) :: list
  def bounding_box(point, radius_in_m) do
    Calculator.bounding_box(point, radius_in_m)
  end

  @doc """
  Calculates a bounding box for a list of points.

  Returns an array with 2 points (list format). The bottom left (southwest) point,
  and the top-right (northeast) one.

  ## Examples

      iex> berlin = [52.5075419, 13.4251364]
      iex> london = [51.5286416, -0.1015987]
      iex> paris = [48.8588589, 2.3475569]
      iex> Geocalc.bounding_box_for_points([berlin, london, paris])
      [[48.8588589, -0.1015987], [52.5075419, 13.4251364]]

  """
  @spec bounding_box_for_points(list) :: list
  def bounding_box_for_points(points) do
    Calculator.bounding_box_for_points(points)
  end

  @doc """
  Extend the bounds to contain the given bounds.

  Returns an array with 2 points (list format). The bottom left (southwest) point,
  and the top-right (northeast) one.

  ## Examples

      iex> berlin = [52.5075419, 13.4251364]
      iex> london = [51.5286416, -0.1015987]
      iex> Geocalc.extend_bounding_box([berlin, berlin], [london, london])
      [[51.5286416, -0.1015987], [52.5075419, 13.4251364]]

  """
  @spec extend_bounding_box(list, list) :: list
  def extend_bounding_box(bounding_box_1, bounding_box_2) do
    Calculator.extend_bounding_box(bounding_box_1, bounding_box_2)
  end

  @doc """
  Returns `true` if the bounding box contains the given point.

  ## Examples

      iex> germany = [[47.27, 5.87], [55.1, 15.04]]
      iex> berlin = [52.5075419, 13.4251364]
      iex> Geocalc.contains_point?(germany, berlin)
      true

  """
  @spec contains_point?(list, Point.t()) :: boolean
  def contains_point?(bounding_box, point) do
    Calculator.contains_point?(bounding_box, point)
  end

  @doc """
  Returns `true` if the bounding box intersects the given bounds.

  Two bounds intersect if they have at least one point in common.

  ## Examples

      iex> germany = [[47.27, 5.87], [55.1, 15.04]]
      iex> poland = [[49.0, 14.12], [55.03, 24.15]]
      iex> Geocalc.intersects_bounding_box?(germany, poland)
      true

  """
  @spec intersects_bounding_box?(list, list) :: boolean
  def intersects_bounding_box?(bounding_box_1, bounding_box_2) do
    Calculator.intersects_bounding_box?(bounding_box_1, bounding_box_2)
  end

  @doc """
  Returns `true` if the bounding box overlaps the given bounds.

  Two bounds overlap if their intersection is an area.

  ## Examples

      iex> germany = [[47.27, 5.87], [55.1, 15.04]]
      iex> berlin_suburbs = [[52.338261, 13.08835], [52.67551, 13.76116]]
      iex> Geocalc.overlaps_bounding_box?(germany, berlin_suburbs)
      true

  """
  @spec overlaps_bounding_box?(list, list) :: boolean
  def overlaps_bounding_box?(bounding_box_1, bounding_box_2) do
    Calculator.overlaps_bounding_box?(bounding_box_1, bounding_box_2)
  end

  @doc """
  Compute the geographic center (aka geographic midpoint, center of gravity)
  for an array of geocoded objects and/or [lat,lon] arrays (can be mixed).

  Any objects missing coordinates are ignored. Follows the procedure
  documented at http://www.geomidpoint.com/calculation.html.

  ## Examples

      iex> point_1 = [0, 0]
      iex> point_2 = [0, 3]
      iex> Geocalc.geographic_center([point_1, point_2])
      [0.0, 1.5]

  """
  @spec geographic_center(list) :: Point.t()
  def geographic_center(points) do
    Calculator.geographic_center(points)
  end

  @doc """
  Converts radians to degrees.

  Returns degrees.

  ## Examples

      iex> Geocalc.radians_to_degrees(2.5075419)
      143.67156782221554

      iex> Geocalc.radians_to_degrees(-0.1015987)
      -5.821176714015797

  """
  @spec radians_to_degrees(number) :: number
  def radians_to_degrees(radians) do
    Calculator.radians_to_degrees(radians)
  end

  @doc """
  Converts degrees to radians.

  Returns radians.

  ## Examples

      iex> Geocalc.degrees_to_radians(143.67156782221554)
      2.5075419

      iex> Geocalc.degrees_to_radians(-10.735322818996854)
      -0.18736672945597435

  """
  @spec degrees_to_radians(number) :: number
  def degrees_to_radians(degrees) do
    Calculator.degrees_to_radians(degrees)
  end

  @doc """
  Calculates maximum latitude reached when travelling on a great circle on given
  bearing from the point (Clairaut's formula). Negate the result for the
  minimum latitude (in the Southern hemisphere).

  The maximum latitude is independent of longitude; it will be the same for all
  points on a given latitude.

  Returns radians.

  ## Examples

      iex> berlin = [52.5075419, 13.4251364]
      iex> paris = [48.8588589, 2.3475569]
      iex> bearing = Geocalc.bearing(berlin, paris)
      iex> Geocalc.max_latitude(berlin, bearing)
      55.953467429882835

  """
  @spec max_latitude(Point.t(), number) :: number
  def max_latitude(point, bearing) do
    Calculator.max_latitude(point, bearing)
  end

  @doc """
  Compute distance from the point to great circle defined by start-point
  and end-point.

  Returns distance in meters.

  ## Examples

      iex> berlin = [52.5075419, 13.4251364]
      iex> london = [51.5286416, -0.1015987]
      iex> paris = [48.8588589, 2.3475569]
      iex> Geocalc.cross_track_distance_to(berlin, london, paris)
      -877680.2992295175

  """
  @spec cross_track_distance_to(Point.t(), Point.t(), Point.t()) :: number
  def cross_track_distance_to(point, path_start_point, path_end_point) do
    Calculator.cross_track_distance_to(point, path_start_point, path_end_point)
  end

  @doc """
  Calculates how far the point is along a path from from start-point, heading
  towards end-point.

  That is, if a perpendicular is drawn from the point to the (great circle)
  path, the along-track distance is the distance from the start point to where
  the perpendicular crosses the path.

  ## Examples

      iex> berlin = [52.5075419, 13.4251364]
      iex> london = [51.5286416, -0.1015987]
      iex> paris = [48.8588589, 2.3475569]
      iex> Geocalc.along_track_distance_to(berlin, london, paris)
      310412.6031976226

  """
  @spec along_track_distance_to(Point.t(), Point.t(), Point.t()) :: number
  def along_track_distance_to(point, path_start_point, path_end_point) do
    Calculator.along_track_distance_to(point, path_start_point, path_end_point)
  end

  @doc """
  Calculates the pair of meridians at which a great circle defined by two points
  crosses the given latitude.

  Returns longitudes.

  ## Examples

      iex> berlin = [52.5075419, 13.4251364]
      iex> paris = [48.8588589, 2.3475569]
      iex> Geocalc.crossing_parallels(berlin, paris, 12.3456)
      {:ok, 123.179463369946, -39.81144878508576}

      iex> point_1 = %{lat: 0, lng: 0}
      iex> point_2 = %{lat: -180, lng: -90}
      iex> latitude = 45.0
      iex> Geocalc.crossing_parallels(point_1, point_2, latitude)
      {:error, "Not found"}

  """
  @spec crossing_parallels(Point.t(), Point.t(), number) :: tuple
  def crossing_parallels(point_1, path_2, latitude) do
    Calculator.crossing_parallels(point_1, path_2, latitude)
  end

  @doc """
  Check if a `point` is inside `area`.

  Returns true if inside area, false if not.

  ## Examples

      iex> area = %Geocalc.Shape.Circle{latitude: 48.856614, longitude: 2.3522219, radius: 1000}
      iex> point = %{lat: 48.856612, lng: 2.3522217}
      iex> Geocalc.in_area?(area, point)
      true

  """
  @spec in_area?(Shape.Circle.t() | Shape.Rectangle.t() | Shape.Ellipse.t(), Point.t()) ::
          boolean
  def in_area?(area, point) do
    Area.point_in_area?(area, point)
  end

  @doc """
  Check if a `point` is outside `area`.

  Returns true if outside area, false if not

  ## Examples

      iex> area = %Geocalc.Shape.Circle{latitude: 48.856614, longitude: 2.3522219, radius: 10}
      iex> point = %{lat: 48.856418, lng: 2.365871}
      iex> Geocalc.outside_area?(area, point)
      true

  """
  @spec outside_area?(Shape.Circle.t() | Shape.Rectangle.t() | Shape.Ellipse.t(), Point.t()) ::
          boolean
  def outside_area?(area, point) do
    Area.point_outside_area?(area, point)
  end

  @doc """
  Check if a `point` is at the border of `area`.

  Returns true if at border, false if not.

  ## Examples

      iex> area = %Geocalc.Shape.Circle{latitude: 48.856614, longitude: 2.3522219, radius: 1000}
      iex> point = %{lat: 48.856418, lng: 2.365871}
      iex> Geocalc.at_area_border?(area, point)
      true

  """
  @spec at_area_border?(Shape.Circle.t() | Shape.Rectangle.t() | Shape.Ellipse.t(), Point.t()) ::
          boolean
  def at_area_border?(area, point) do
    Area.point_at_area_border?(area, point)
  end

  @doc """
  Check if a `point` at the center point of `area`.

  Returns true if at center point, false if not

  ## Examples

      iex> area = %Geocalc.Shape.Circle{latitude: 48.856614, longitude: 2.3522219, radius: 100}
      iex> point = %{lat: 48.856614, lng: 2.3522219}
      iex> Geocalc.at_center_point?(area, point)
      true

  """
  @spec at_center_point?(Shape.Circle.t() | Shape.Rectangle.t() | Shape.Ellipse.t(), Point.t()) ::
          boolean
  def at_center_point?(area, point) do
    Area.point_at_center_point?(area, point)
  end

  @doc """
  Calculate the given `area` surface.

  Returns `area` surface in square meters.

  ## Examples

      iex> area = %Geocalc.Shape.Circle{latitude: 48.856614, longitude: 2.3522219, radius: 1000}
      iex> Geocalc.area_size(area)
      3141592.653589793

  """
  @spec area_size(Shape.Circle.t() | Shape.Rectangle.t() | Shape.Ellipse.t()) :: non_neg_integer
  def area_size(area) do
    Area.area_size(area)
  end
end
